src/runtime_src/hip/core/module.cpp

```diff
@@ -8,50 +8,37 @@
 
 namespace xrt::core::hip {
 void
-module::
+module_xclbin::
 create_hw_context()
 {
-  auto xrt_dev = m_ctx->get_xrt_device();
+  auto xrt_dev = get_context()->get_xrt_device();
   auto uuid = xrt_dev.register_xclbin(m_xclbin);
   m_hw_ctx = xrt::hw_context{xrt_dev, uuid};
 }
 
-module::
-module(std::shared_ptr<context> ctx, const std::string& file_name)
-  : m_ctx{std::move(ctx)}
+module_xclbin::
+module_xclbin(std::shared_ptr<context> ctx, const std::string& file_name)
+  : module{ctx, true}
 {
   m_xclbin = xrt::xclbin{file_name};
   create_hw_context();
 }
 
-module::
-module(std::shared_ptr<context> ctx, const void* image)
-  : m_ctx{std::move(ctx)}
+module_elf::
+module_elf(module_xclbin* xclbin_module, const std::string& file_name)
+  : module(xclbin_module->get_context(), false)
+  , m_xclbin_module(xclbin_module)
 {
-  // we trust pointer sent by application and treat
-  // it as xclbin data. Application can crash/seg fault
-  // when improper data is passed
-  m_xclbin = xrt::xclbin{static_cast<const axlf*>(image)};
-  create_hw_context();
-}
-
-xrt::kernel
-module::
-create_kernel(std::string& name)
-{
-  return xrt::kernel{m_hw_ctx, name};
+  m_elf = xrt::elf{file_name};
+  m_module = xrt::module{m_elf};
 }
 
 function::
-function(module_handle mod_hdl, std::string&& name)
-  : m_module{static_cast<module*>(mod_hdl)}
-  , m_func_name{std::move(name)}
+function(module_xclbin* xclbin_mod_hdl, const xrt::module& xrt_module, const std::string& name)
+  : m_module(xclbin_mod_hdl)
+  , m_func_name{name}
 {
-  if (!module_cache.count(mod_hdl))
-    throw xrt_core::system_error(hipErrorInvalidResourceHandle, "module not available");
-
-
-  m_kernel = m_module->create_kernel(m_func_name);
+  m_kernel = xrt::ext::kernel{m_module->get_hw_context(), xrt_module, name};
```

Comment:
This should be done in initializer list line 39.
, m_kernel{xrt::ext::kernel{m_module->get_hw_context(), xrt_module, name}};

I think #8198 unfortunately implicitly removed the move-ctor from xrt::kernel, which we should add back in, but this shouldn't matter for you, it's just an unnecessary reference count update.